File: apps/dash-live-model-training/demo_utils.py
from dash import Dash, html, dcc, Input, Output, State, callback, callback_context
import pandas as pd
import pathlib

# get relative data folder
PATH = pathlib.Path(__file__).parent
DATA_PATH = PATH.joinpath("data").resolve()


def demo_explanation(demo_mode):
    if demo_mode:
        # Markdown files
        with open(PATH.joinpath("demo.md"), "r") as file:
            demo_md = file.read()

        return html.Div(
            html.Div([dcc.Markdown(demo_md, className="markdown")]),
            style={"margin": "10px"},
        )


def demo_callbacks(app, demo_mode):
    if demo_mode:

        @app.server.before_first_request
        def load_demo_run_logs():
            global data_dict, demo_md

            names = [
                "step",
                "train accuracy",
                "val accuracy",
                "train cross entropy",
                "val cross entropy",
            ]

            data_dict = {
                "softmax": {
                    "cifar": pd.read_csv(
                        DATA_PATH.joinpath("cifar_softmax_run_log.csv"), names=names
                    ),
                    "mnist": pd.read_csv(
                        DATA_PATH.joinpath("mnist_softmax_run_log.csv"), names=names
                    ),
                    "fashion": pd.read_csv(
                        DATA_PATH.joinpath("fashion_softmax_run_log.csv"), names=names
                    ),
                },
                "cnn": {
                    "cifar": pd.read_csv(
                        DATA_PATH.joinpath("cifar_cnn_run_log.csv"), names=names
                    ),
                    "mnist": pd.read_csv(
                        DATA_PATH.joinpath("mnist_cnn_run_log.csv"), names=names
                    ),
                    "fashion": pd.read_csv(
                        DATA_PATH.joinpath("fashion_cnn_run_log.csv"), names=names
                    ),
                },
            }

        @app.callback(
            Output("storage-simulated-run", "data"),
            [Input("interval-simulated-step", "n_intervals")],
            [
                State("dropdown-demo-dataset", "value"),
                State("dropdown-simulation-model", "value"),
            ],
        )
        def simulate_run(n_intervals, demo_dataset, simulation_model):
            if simulation_model and demo_dataset and n_intervals > 0:
                step = n_intervals * 5
                run_logs = data_dict[simulation_model][demo_dataset]

                run_below_steps = run_logs[run_logs["step"] <= step]
                json = run_below_steps.to_json(orient="split")

                return json

        @app.callback(
            Output("interval-simulated-step", "n_intervals"),
            [
                Input("dropdown-demo-dataset", "value"),
                Input("dropdown-simulation-model", "value"),
            ],
        )
        def reset_interval_simulated_step(*_):
            return 0

        @app.callback(
            Output("run-log-storage", "data"),
            [Input("interval-log-update", "n_intervals")],
            [State("storage-simulated-run", "data")],
        )
        def get_run_log(_, simulated_run):
            if simulate_run:
                return simulated_run

        @app.callback(
            Output("div-total-step-count", "children"),
            [Input("dropdown-demo-dataset", "value")],
        )
        def total_step_count(dataset_name):
            if dataset_name is not None:
                dataset = data_dict["softmax"][dataset_name]
                return html.H6(
                    f"Total Steps: {dataset['step'].iloc[-1]}",
                    style={"margin-top": "3px", "float": "right"},
                )

File: apps/dashr-live-model-training/utils/demo_utils.R
library(dashCoreComponents)
library(dashHtmlComponents)
library(data.table)
library(jsonlite)

demo_callbacks <- function(app, demo_mode){
  if(demo_mode){
    
    demodf <- c("cifar","mnist","fashion")
    simmodels <- c("softmax","cnn")
    
    setwd("C:/Users/MatthiasL/Desktop/DATA/Plotly/RApps/dash-live-model-training-r")
    
    namess <- c("step", "train accuracy", "val accuracy", "train cross entropy", "val cross entropy")
    data_dict <- list(
      "softmax" = list("cifar" = fread('./data/cifar_softmax_run_log.csv',col.names=namess),
                       "mnist" = fread('./data/mnist_softmax_run_log.csv',col.names=namess),
                       "fashion" = fread('./data/fashion_softmax_run_log.csv',col.names=namess)),
      
      "cnn" = list("cifar" = fread('./data/cifar_cnn_run_log.csv',col.names=namess),
                   "mnist" = fread('./data/mnist_cnn_run_log.csv',col.names=namess),
                   "fashion" = fread('./data/fashion_cnn_run_log.csv',col.names=namess))
    )
    
    app$callback(
      output(id="storage-simulated-run","data"),
      list(input("interval-simulated-step","n_intervals"),
           state("dropdown-demo-dataset","value"),
           state("dropdown-simulation-model","value")),
      
      simulate_run <- function(n_intervals, demo_dataset, simulation_model) {
        if((is.element(demo_dataset ,demodf)) &&
           (is.element(simulation_model, simmodels))
           && n_intervals > 0){
          steps <- n_intervals * 5
          run_logs <- data_dict[[simulation_model]][[demo_dataset]]
          run_below_steps <- subset(run_logs, step <= steps)
          json <- toJSON(run_below_steps)
          return(json)
        }
      }
    )
    
    app$callback(
      output("interval-simulated-step","n_intervals"),
      list(input("dropdown-demo-dataset","value"),
           input("dropdown-simulation-model","value")),
      reset_interval_simulated_step <- function(demo_ds,demo_sm){
        return(0)
      }
    )
    
    
    app$callback(
      output("run-log-storage","data"),
      list(input("interval-log-update","n_intervals"),
           state("storage-simulated-run","data")),
      
      get_run_log <- function(n_intervs, simulated_run){
        if(!is.null(simulate_run)){
          return(simulated_run)
        }
      }
    )
    
    #Total steps counter
    app$callback(
      output("div-total-step-count","children"),
      list(input("dropdown-demo-dataset","value")),
      
      total_step_count <- function(dataset_name){
        if(!is.null(dataset_name)){
          dataset <- data_dict$softmax[[dataset_name]]
          return(htmlH6(paste0("Total steps: ",dataset$step[length(dataset$step)]),
                        style= list("margin-top" = "3px",
                                    "float" = "right")
          ))
        }
      }
    )
  }
}
